engines/crab/game.cpp

```diff
@@ -0,0 +1,893 @@
+			loadStr(loc, "loc_id", node);
+			loadLevel(loc);
+
+			if (nodeValid("level", node))
+				_level.loadState(node->first_node("level"));
+
+			_gem._per.Cache(_info, _level.playerId(), _level);
+
+			Common::String playtime;
+			loadStr(playtime, "time", node);
+			_clock.start(playtime);
+		}
+	}
+}
+//------------------------------------------------------------------------
+// Purpose: Write game state to file
+//------------------------------------------------------------------------
+void Game::saveState(Common::SeekableWriteStream *stream) {
+	rapidxml::xml_document<char> doc;
+
+	// xml declaration
+	rapidxml::xml_node<char> *decl = doc.allocate_node(rapidxml::node_declaration);
+	decl->append_attribute(doc.allocate_attribute("version", "1.0"));
+	decl->append_attribute(doc.allocate_attribute("encoding", "utf-8"));
+	doc.append_node(decl);
+
+	// root node
+	rapidxml::xml_node<char> *root = doc.allocate_node(rapidxml::node_element, "save");
+	doc.append_node(root);
+
+	// Save location id
+	Common::String loc = _info.curLocID();
+	root->append_attribute(doc.allocate_attribute("loc_id", loc.c_str()));
+
+	// Save location name
+	Common::String loc_name = _info.curLocName();
+	root->append_attribute(doc.allocate_attribute("loc_name", loc_name.c_str()));
+
+	// Save player character name
+	Common::String char_name;
+	if (_info.personValid(_level.playerId()))
+		char_name = _info.personGet(_level.playerId())._name;
+	root->append_attribute(doc.allocate_attribute("char_name", char_name.c_str()));
+
+	// Difficulty
+	Common::String diff = "Normal";
+	if (_info.ironMan())
+		diff = "Iron Man";
+	root->append_attribute(doc.allocate_attribute("diff", diff.c_str()));
+
+	// Save file used if iron man
+	root->append_attribute(doc.allocate_attribute("file", _savefile._ironman.c_str()));
+
+	// Preview image used
+	root->append_attribute(doc.allocate_attribute("preview", _level._previewPath.c_str()));
+
+	// Time played
+	Common::String playtime = _clock.getTime();
+	root->append_attribute(doc.allocate_attribute("time", playtime.c_str()));
+
+	rapidxml::xml_node<char> *child_gem = doc.allocate_node(rapidxml::node_element, "events");
+	_gem.saveState(doc, child_gem);
+	root->append_node(child_gem);
+
+	rapidxml::xml_node<char> *child_info = doc.allocate_node(rapidxml::node_element, "info");
+	_info.saveState(doc, child_info);
+	root->append_node(child_info);
+
+	rapidxml::xml_node<char> *child_map = doc.allocate_node(rapidxml::node_element, "map");
+	_map.saveState(doc, child_map);
+	root->append_node(child_map);
+
+	rapidxml::xml_node<char> *child_level = doc.allocate_node(rapidxml::node_element, "level");
+	_level.saveState(doc, child_level);
+	root->append_node(child_level);
+
+	Common::String xml_as_string;
+	rapidxml::print(Crab::backInserter(xml_as_string), doc);
+
+	stream->writeString(xml_as_string);
+
+#if 0
+	Common::String fullpath = FullPath(filename);
+
+	// We don't check for duplicates for auto-saves and iron man saves
```

Comment:
We also have autosaves in ScummVM. Maybe it is worth supporting?

Reply:
Autosaves are already supported and are working. Dead code. Removed.